// eas-cli/build/commands/submit/upload-to-asc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const core_1 = require("@oclif/core");
const fs_extra_1 = tslib_1.__importDefault(require("fs-extra"));
const jsonwebtoken_1 = tslib_1.__importDefault(require("jsonwebtoken"));
const node_fetch_1 = tslib_1.__importDefault(require("node-fetch"));
const path_1 = tslib_1.__importDefault(require("path"));
const promises_1 = require("timers/promises");
const zod_1 = require("zod");
const EasCommand_1 = tslib_1.__importDefault(require("../../commandUtils/EasCommand"));
const log_1 = tslib_1.__importDefault(require("../../log"));
const AscApiClient_1 = require("../../submit/ios/AscApiClient");
class SubmitUploadToAsc extends EasCommand_1.default {
    static hidden = true;
    static flags = {
        path: core_1.Flags.string({
            description: 'Path to the IPA file',
            required: true,
        }),
        key: core_1.Flags.string({
            description: 'Path to the ASC API Key JSON file',
            required: true,
        }),
        'app-id': core_1.Flags.string({
            description: 'App Store Connect App ID (e.g. 1491144534)',
            required: true,
        }),
        'bundle-version': core_1.Flags.string({
            description: 'CFBundleVersion (Build Version, e.g. 13)',
            required: true,
        }),
        'bundle-short-version': core_1.Flags.string({
            description: 'CFBundleShortVersionString (Marketing Version, e.g. 1.0.0)',
            required: true,
        }),
    };
    async runAsync() {
        const { flags } = await this.parse(SubmitUploadToAsc);
        const ipaPath = path_1.default.resolve(flags.path);
        if (!(await fs_extra_1.default.pathExists(ipaPath))) {
            throw new Error(`IPA file not found: ${ipaPath}`);
        }
        const fileSize = (await fs_extra_1.default.stat(ipaPath)).size;
        const fileName = path_1.default.basename(ipaPath);
        const keyPath = path_1.default.resolve(flags.key);
        if (!(await fs_extra_1.default.pathExists(keyPath))) {
            throw new Error(`Key file not found: ${keyPath}`);
        }
        const keyJson = await fs_extra_1.default.readJson(keyPath);
        const apiKey = zod_1.z
            .object({
            issuer_id: zod_1.z.string(),
            key_id: zod_1.z.string(),
            key: zod_1.z.string(),
        })
            .parse(keyJson);
        log_1.default.log('Generating JWT...');
        const token = jsonwebtoken_1.default.sign({}, apiKey.key, {
            algorithm: 'ES256',
            issuer: apiKey.issuer_id,
            expiresIn: '20m',
            audience: 'appstoreconnect-v1',
            header: {
                kid: apiKey.key_id,
            },
        });
        const client = new AscApiClient_1.AscApiClient({ token });
        log_1.default.log('Reading App information...');
        const appResponse = await client.getAsync('/v1/apps/:id', { 'fields[apps]': ['bundleId', 'name'] }, { id: flags['app-id'] });
        log_1.default.log(`Uploading Build to "${appResponse.data.attributes.name}" (${appResponse.data.attributes.bundleId})...`);
        log_1.default.log('Creating Build Upload...');
        const buildUploadResponse = await client.postAsync('/v1/buildUploads', {
            data: {
                type: 'buildUploads',
                attributes: {
                    platform: 'IOS',
                    cfBundleShortVersionString: flags['bundle-short-version'],
                    cfBundleVersion: flags['bundle-version'],
                },
                relationships: {
                    app: {
                        data: {
                            type: 'apps',
                            id: flags['app-id'],
                        },
                    },
                },
            },
        });
        const buildUploadId = buildUploadResponse.data.id;
        log_1.default.log(`Build Upload initialized (ID: ${buildUploadId}). Preparing IPA upload...`);
        const buildFileResponse = await client.postAsync('/v1/buildUploadFiles', {
            data: {
                type: 'buildUploadFiles',
                attributes: {
                    assetType: 'ASSET',
                    fileName,
                    fileSize,
                    uti: 'com.apple.ipa',
                },
                relationships: {
                    buildUpload: {
                        data: {
                            type: 'buildUploads',
                            id: buildUploadId,
                        },
                    },
                },
            },
        });
        log_1.default.log(`IPA Upload initialized (ID: ${buildFileResponse.data.id}). Uploading IPA...`);
        await uploadChunksAsync({
            uploadOperations: buildFileResponse.data.attributes.uploadOperations,
            ipaPath,
        });
        log_1.default.log('Committing upload...');
        await client.patchAsync(`/v1/buildUploadFiles/:id`, {
            data: {
                type: 'buildUploadFiles',
                id: buildFileResponse.data.id,
                attributes: {
                    uploaded: true,
                },
            },
        }, {
            id: buildFileResponse.data.id,
        });
        log_1.default.log('Checking upload file status...');
        const waitingForFileStartedAt = performance.now();
        while (performance.now() - waitingForFileStartedAt < 60 * 1000 /* 60 seconds */) {
            const buildFileStatusResponse = await client.getAsync(`/v1/buildUploadFiles/:id`, { 'fields[buildUploadFiles]': ['assetDeliveryState'] }, { id: buildFileResponse.data.id });
            if (buildFileStatusResponse.data.attributes.assetDeliveryState.state === 'AWAITING_UPLOAD') {
                log_1.default.log(`Waiting for file upload to finish processing... (state = ${buildFileStatusResponse.data.attributes.assetDeliveryState.state})`);
                await (0, promises_1.setTimeout)(2000);
                continue;
            }
            const { errors = [], warnings = [] } = buildFileStatusResponse.data.attributes.assetDeliveryState;
            if (warnings.length > 0) {
                log_1.default.warn(`Warnings:\n- ${warnings.map(w => `${w.description} (${w.code})`).join('\n- ')}\n`);
            }
            if (errors.length > 0) {
                log_1.default.error(`Errors:\n- ${errors.map(e => `${e.description} (${e.code})`).join('\n- ')}\n`);
            }
            if (buildFileStatusResponse.data.attributes.assetDeliveryState.state === 'FAILED') {
                throw new Error(`File upload (ID: ${buildFileResponse.data.id}) failed.`);
            }
            else if (buildFileStatusResponse.data.attributes.assetDeliveryState.state === 'COMPLETE') {
                log_1.default.log(`File upload (ID: ${buildFileResponse.data.id}) complete!`);
            }
            else if (buildFileStatusResponse.data.attributes.assetDeliveryState.state === 'UPLOAD_COMPLETE') {
                log_1.default.log(`File upload (ID: ${buildFileResponse.data.id}) finished!`);
            }
            break;
        }
        log_1.default.log('Checking build upload status...');
        const waitingForBuildStartedAt = performance.now();
        while (performance.now() - waitingForBuildStartedAt < 60 * 1000 /* 60 seconds */) {
            const buildUploadStatusResponse = await client.getAsync(`/v1/buildUploads/:id`, { 'fields[buildUploads]': ['state', 'build'], include: ['build'] }, { id: buildUploadId });
            if (buildUploadStatusResponse.data.attributes.state.state === 'AWAITING_UPLOAD' ||
                buildUploadStatusResponse.data.attributes.state.state === 'PROCESSING') {
                log_1.default.log(`Waiting for build upload to finish... (status = ${buildUploadStatusResponse.data.attributes.state.state})`);
                await (0, promises_1.setTimeout)(2000);
                continue;
            }
            log_1.default.log('\n');
            const { errors = [], warnings = [], infos = [], } = buildUploadStatusResponse.data.attributes.state;
            if (infos.length > 0) {
                log_1.default.log(`Infos:\n- ${infos.map(i => `${i.description} (${i.code})`).join('\n- ')}\n`);
            }
            if (warnings.length > 0) {
                log_1.default.warn(`Warnings:\n- ${warnings.map(w => `${w.description} (${w.code})`).join('\n- ')}\n`);
            }
            if (errors.length > 0) {
                log_1.default.error(`Errors:\n- ${errors.map(e => `${e.description} (${e.code})`).join('\n- ')}\n`);
            }
            if (buildUploadStatusResponse.data.attributes.state.state === 'FAILED') {
                throw new Error(`Build upload (ID: ${buildUploadId}) failed.`);
            }
            else if (buildUploadStatusResponse.data.attributes.state.state === 'COMPLETE') {
                log_1.default.log(`Build upload (ID: ${buildUploadId}) complete!`);
            }
            break;
        }
    }
}
exports.default = SubmitUploadToAsc;
async function uploadChunksAsync({ uploadOperations, ipaPath, }) {
    const fd = await fs_extra_1.default.open(ipaPath, 'r');
    try {
        const promises = uploadOperations.map(async (op, index) => {
            log_1.default.log(`  Uploading chunk ${index + 1}/${uploadOperations.length} (offset: ${op.offset}, length: ${op.length})...`);
            const buffer = new Uint8Array(op.length);
            await fs_extra_1.default.read(fd, buffer, 0, op.length, op.offset);
            const headers = {
                'Content-Type': 'application/octet-stream',
            };
            for (const { name, value } of op.requestHeaders) {
                headers[name] = value;
            }
            const response = await (0, node_fetch_1.default)(op.url, {
                method: op.method,
                headers,
                body: buffer,
            });
            if (!response.ok) {
                const text = await response.text();
                throw new Error(`Failed to upload chunk ${index + 1}: ${response.status} ${response.statusText}\n${text}`);
            }
            log_1.default.log(`  Chunk ${index + 1}/${uploadOperations.length} uploaded.`);
        });
        await Promise.all(promises);
    }
    finally {
        await fs_extra_1.default.close(fd);
    }
}
