// eas-cli/build/submit/ios/AscApiClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AscApiClient = void 0;
const tslib_1 = require("tslib");
const results_1 = require("@expo/results");
const node_fetch_1 = tslib_1.__importDefault(require("node-fetch"));
const zod_1 = require("zod");
const log_1 = tslib_1.__importDefault(require("../../log"));
const GetApi = {
    '/v1/apps/:id': {
        path: zod_1.z.object({
            id: zod_1.z.string(),
        }),
        request: zod_1.z.object({
            'fields[apps]': zod_1.z.array(zod_1.z.enum(['bundleId', 'name'])).refine(opts => {
                // Let's say we currently require fetching these two and nothing else.
                return opts.includes('bundleId') && opts.includes('name');
            }),
        }),
        response: zod_1.z.object({
            data: zod_1.z.object({
                type: zod_1.z.literal('apps'),
                id: zod_1.z.string(),
                attributes: zod_1.z.object({
                    bundleId: zod_1.z.string(),
                    name: zod_1.z.string(),
                }),
            }),
        }),
    },
    '/v1/buildUploadFiles/:id': {
        path: zod_1.z.object({
            id: zod_1.z.string(),
        }),
        request: zod_1.z.object({
            'fields[buildUploadFiles]': zod_1.z.array(zod_1.z.enum(['assetDeliveryState'])).refine(opts => {
                return opts.includes('assetDeliveryState');
            }),
        }),
        response: zod_1.z.object({
            data: zod_1.z.object({
                type: zod_1.z.literal('buildUploadFiles'),
                id: zod_1.z.string(),
                attributes: zod_1.z.object({
                    // https://developer.apple.com/documentation/appstoreconnectapi/appmediaassetstate
                    assetDeliveryState: zod_1.z.object({
                        state: zod_1.z.enum(['AWAITING_UPLOAD', 'UPLOAD_COMPLETE', 'COMPLETE', 'FAILED']),
                        errors: zod_1.z.array(zod_1.z.object({ code: zod_1.z.string(), description: zod_1.z.string() })).optional(),
                        warnings: zod_1.z.array(zod_1.z.object({ code: zod_1.z.string(), description: zod_1.z.string() })).optional(),
                    }),
                }),
            }),
        }),
    },
    '/v1/buildUploads/:id': {
        path: zod_1.z.object({
            id: zod_1.z.string(),
        }),
        request: zod_1.z.object({
            'fields[buildUploads]': zod_1.z.array(zod_1.z.enum(['build', 'state'])).refine(opts => {
                return opts.includes('build') && opts.includes('state');
            }),
            include: zod_1.z.array(zod_1.z.enum(['build'])).refine(opts => {
                return opts.includes('build');
            }),
        }),
        // https://developer.apple.com/documentation/appstoreconnectapi/builduploadresponse
        response: zod_1.z.object({
            data: zod_1.z.object({
                type: zod_1.z.literal('buildUploads'),
                id: zod_1.z.string(),
                attributes: zod_1.z.object({
                    // https://developer.apple.com/documentation/appstoreconnectapi/buildupload/attributes-data.dictionary/state-data.dictionary
                    state: zod_1.z.object({
                        state: zod_1.z.enum(['AWAITING_UPLOAD', 'PROCESSING', 'COMPLETE', 'FAILED']),
                        infos: zod_1.z.array(zod_1.z.object({ code: zod_1.z.string(), description: zod_1.z.string() })).optional(),
                        errors: zod_1.z.array(zod_1.z.object({ code: zod_1.z.string(), description: zod_1.z.string() })).optional(),
                        warnings: zod_1.z.array(zod_1.z.object({ code: zod_1.z.string(), description: zod_1.z.string() })).optional(),
                    }),
                }),
            }),
        }),
    },
};
const PostApi = {
    '/v1/buildUploads': {
        request: zod_1.z.object({
            // https://developer.apple.com/documentation/appstoreconnectapi/builduploadcreaterequest/data-data.dictionary
            data: zod_1.z.object({
                type: zod_1.z.literal('buildUploads'),
                // https://developer.apple.com/documentation/appstoreconnectapi/builduploadcreaterequest/data-data.dictionary/attributes-data.dictionary
                attributes: zod_1.z.object({
                    cfBundleShortVersionString: zod_1.z.string(),
                    cfBundleVersion: zod_1.z.string(),
                    // https://developer.apple.com/documentation/appstoreconnectapi/platform
                    platform: zod_1.z.enum(['IOS', 'MAC_OS', 'TV_OS', 'VISION_OS']),
                }),
                // https://developer.apple.com/documentation/appstoreconnectapi/builduploadcreaterequest/data-data.dictionary/relationships-data.dictionary
                relationships: zod_1.z.object({
                    // https://developer.apple.com/documentation/appstoreconnectapi/builduploadcreaterequest/data-data.dictionary/relationships-data.dictionary/app-data.dictionary
                    app: zod_1.z.object({
                        // https://developer.apple.com/documentation/appstoreconnectapi/builduploadcreaterequest/data-data.dictionary/relationships-data.dictionary/app-data.dictionary/data-data.dictionary
                        data: zod_1.z.object({
                            type: zod_1.z.literal('apps'),
                            id: zod_1.z.string(),
                        }),
                    }),
                }),
            }),
        }),
        // https://developer.apple.com/documentation/appstoreconnectapi/builduploadresponse
        response: zod_1.z.object({
            // https://developer.apple.com/documentation/appstoreconnectapi/buildupload
            data: zod_1.z.object({
                id: zod_1.z.string(),
                type: zod_1.z.literal('buildUploads'),
            }),
        }),
    },
    '/v1/buildUploadFiles': {
        // https://developer.apple.com/documentation/appstoreconnectapi/builduploadfilecreaterequest
        request: zod_1.z.object({
            // https://developer.apple.com/documentation/appstoreconnectapi/builduploadfilecreaterequest/data-data.dictionary
            data: zod_1.z.object({
                type: zod_1.z.literal('buildUploadFiles'),
                // https://developer.apple.com/documentation/appstoreconnectapi/builduploadfilecreaterequest/data-data.dictionary/attributes-data.dictionary
                attributes: zod_1.z.object({
                    assetType: zod_1.z.enum(['ASSET', 'ASSET_DESCRIPTION', 'ASSET_SPI']),
                    fileName: zod_1.z.string(),
                    fileSize: zod_1.z.number().min(1).max(9007199254740991),
                    uti: zod_1.z.enum([
                        'com.apple.binary-property-list',
                        'com.apple.ipa',
                        'com.apple.pkg',
                        'com.apple.xml-property-list',
                        'com.pkware.zip-archive',
                    ]),
                }),
                // https://developer.apple.com/documentation/appstoreconnectapi/builduploadfilecreaterequest/data-data.dictionary/relationships-data.dictionary
                relationships: zod_1.z.object({
                    // https://developer.apple.com/documentation/appstoreconnectapi/builduploadfilecreaterequest/data-data.dictionary/relationships-data.dictionary/buildupload-data.dictionary
                    buildUpload: zod_1.z.object({
                        // https://developer.apple.com/documentation/appstoreconnectapi/builduploadfilecreaterequest/data-data.dictionary/relationships-data.dictionary/buildupload-data.dictionary/data-data.dictionary
                        data: zod_1.z.object({
                            type: zod_1.z.literal('buildUploads'),
                            id: zod_1.z.string(),
                        }),
                    }),
                }),
            }),
        }),
        // https://developer.apple.com/documentation/appstoreconnectapi/builduploadfileresponse
        response: zod_1.z.object({
            // https://developer.apple.com/documentation/appstoreconnectapi/builduploadfile
            data: zod_1.z.object({
                type: zod_1.z.literal('buildUploadFiles'),
                id: zod_1.z.string(),
                // https://developer.apple.com/documentation/appstoreconnectapi/builduploadfile/attributes-data.dictionary
                attributes: zod_1.z.object({
                    // https://developer.apple.com/documentation/appstoreconnectapi/appmediaassetstate
                    uploadOperations: zod_1.z.array(
                    // https://developer.apple.com/documentation/appstoreconnectapi/deliveryfileuploadoperation
                    zod_1.z.object({
                        length: zod_1.z.number().min(1).max(9007199254740991),
                        method: zod_1.z.string(),
                        offset: zod_1.z.number().min(0).max(9007199254740991),
                        partNumber: zod_1.z.number().min(1).max(9007199254740991),
                        requestHeaders: zod_1.z.array(zod_1.z.object({
                            name: zod_1.z.string(),
                            value: zod_1.z.string(),
                        })),
                        url: zod_1.z.string(),
                    })),
                }),
            }),
        }),
    },
};
const PatchApi = {
    // https://developer.apple.com/documentation/appstoreconnectapi/patch-v1-builduploadfiles-_id_
    '/v1/buildUploadFiles/:id': {
        path: zod_1.z.object({
            id: zod_1.z.string(),
        }),
        request: zod_1.z.object({
            // https://developer.apple.com/documentation/appstoreconnectapi/builduploadfileupdaterequest/data-data.dictionary
            data: zod_1.z.object({
                id: zod_1.z.string(),
                type: zod_1.z.literal('buildUploadFiles'),
                // https://developer.apple.com/documentation/appstoreconnectapi/builduploadfileupdaterequest/data-data.dictionary/attributes-data.dictionary
                attributes: zod_1.z.object({
                    uploaded: zod_1.z.boolean(),
                }),
            }),
        }),
        response: zod_1.z.object({
            // https://developer.apple.com/documentation/appstoreconnectapi/builduploadfile
            data: zod_1.z.object({
                type: zod_1.z.literal('buildUploadFiles'),
                id: zod_1.z.string(),
                // https://developer.apple.com/documentation/appstoreconnectapi/builduploadfile/attributes-data.dictionary
                attributes: zod_1.z.object({
                    // https://developer.apple.com/documentation/appstoreconnectapi/appmediaassetstate
                    assetDeliveryState: zod_1.z.object({
                        state: zod_1.z.enum(['AWAITING_UPLOAD', 'UPLOAD_COMPLETE', 'COMPLETE', 'FAILED']),
                        errors: zod_1.z.array(zod_1.z.object({ code: zod_1.z.string(), description: zod_1.z.string() })).optional(),
                        warnings: zod_1.z.array(zod_1.z.object({ code: zod_1.z.string(), description: zod_1.z.string() })).optional(),
                    }),
                }),
            }),
        }),
    },
};
class AscApiClient {
    baseUrl = 'https://api.appstoreconnect.apple.com';
    token;
    constructor({ token }) {
        this.token = token;
    }
    async getAsync(path, body, params) {
        const schema = GetApi[path];
        let effectivePath = path;
        for (const [key, value] of Object.entries(params ?? {})) {
            effectivePath = effectivePath.replace(`:${key}`, String(value));
        }
        return await this.sendRequestAsync({
            method: 'GET',
            path: `${effectivePath}?${new URLSearchParams(body).toString()}`,
            body,
            requestSchema: schema.request,
            responseSchema: schema.response,
        });
    }
    async postAsync(path, body) {
        const schema = PostApi[path];
        return await this.sendRequestAsync({
            method: 'POST',
            path,
            body,
            requestSchema: schema.request,
            responseSchema: schema.response,
        });
    }
    async patchAsync(path, body, params) {
        const schema = PatchApi[path];
        let effectivePath = path;
        for (const [key, value] of Object.entries(params)) {
            effectivePath = effectivePath.replace(`:${key}`, String(value));
        }
        return await this.sendRequestAsync({
            method: 'PATCH',
            path: effectivePath,
            body,
            requestSchema: schema.request,
            responseSchema: schema.response,
        });
    }
    async sendRequestAsync({ path, method, body, requestSchema, responseSchema, }) {
        const url = new URL(path, this.baseUrl).toString();
        const parsedBody = await (0, results_1.asyncResult)((async () => requestSchema.parse(body))());
        if (!parsedBody.ok) {
            throw new Error(`Malformed request to App Store Connect: ${zod_1.z.prettifyError(parsedBody.enforceError())}`);
        }
        const response = await (0, node_fetch_1.default)(url, {
            method,
            headers: {
                'Content-Type': 'application/json',
                Authorization: `Bearer ${this.token}`,
            },
            body: method === 'GET' ? undefined : JSON.stringify(parsedBody.value),
        });
        if (!response.ok) {
            const text = await response.text();
            throw new Error(`Unexpected response (${response.status}) from App Store Connect: ${text}`, {
                cause: response,
            });
        }
        const json = await response.json();
        log_1.default.debug(`Response from App Store Connect: ${JSON.stringify(json, null, 2)}`);
        const parsedResponse = await (0, results_1.asyncResult)((async () => responseSchema.parse(json))());
        if (!parsedResponse.ok) {
            throw new Error(`Malformed response from App Store Connect: ${zod_1.z.prettifyError(parsedResponse.enforceError())}`);
        }
        return parsedResponse.value;
    }
}
exports.AscApiClient = AscApiClient;
